File: internal/integrations/argocd/rbac/glob.go
package rbac

import (
	"errors"
	"fmt"
	"strings"
	"sync"
	"unicode/utf8"
)

// This file is a matcher compatible with github.com/gobwas/glob compiled
// with no separators, which is how Argo CD calls it. With no separators
// `*` and `**` both match any sequence of characters, including "/".
//
// Syntax: `*` and `**` any sequence; `?` one character; `[abc]`, `[a-z]`,
// `[!abc]` character classes; `{a,b}` alternatives, which may nest; `\x`
// escapes x.

type nodeKind int

const (
	nText  nodeKind = iota
	nAny            // ?
	nSuper          // * or **
	nClass          // [...]
)

type node struct {
	kind nodeKind
	text string
	// class
	not    bool
	ranges [][2]rune
}

// globPattern is a compiled pattern: a set of alternative flat sequences.
type globPattern struct {
	alts [][]node
}

var (
	globMu    sync.Mutex
	globCache = map[string]*globPattern{}
	globErrs  = map[string]error{}
)

// globMatch reports whether text matches pattern. An invalid pattern never
// matches, as in Argo CD (which logs the compile error and returns false).
func globMatch(pattern, text string) bool {
	p, err := compileGlob(pattern)
	if err != nil {
		return false
	}
	return p.match(text)
}

func compileGlob(pattern string) (*globPattern, error) {
	globMu.Lock()
	defer globMu.Unlock()
	if p, ok := globCache[pattern]; ok {
		return p, nil
	}
	if err, ok := globErrs[pattern]; ok {
		return nil, err
	}
	if len(globCache)+len(globErrs) > 10000 {
		globCache = map[string]*globPattern{}
		globErrs = map[string]error{}
	}
	p, err := parseGlob(pattern)
	if err != nil {
		globErrs[pattern] = err
		return nil, err
	}
	globCache[pattern] = p
	return p, nil
}

// parseGlob parses the whole pattern.
func parseGlob(pattern string) (*globPattern, error) {
	alts, rest, err := parseSeq(pattern, false)
	if err != nil {
		return nil, err
	}
	if rest != "" {
		return nil, fmt.Errorf("glob %q: unexpected %q", pattern, rest)
	}
	return &globPattern{alts: alts}, nil
}

// parseSeq parses until end of input or, inside braces, until "," or "}".
// It returns every alternative expansion of the parsed sequence.
func parseSeq(s string, inBrace bool) (alts [][]node, rest string, err error) {
	alts = [][]node{{}}
	appendNode := func(n node) {
		for i := range alts {
			alts[i] = append(alts[i], n)
		}
	}
	appendAlts := func(sub [][]node) {
		var out [][]node
		for _, a := range alts {
			for _, b := range sub {
				seq := make([]node, 0, len(a)+len(b))
				seq = append(seq, a...)
				seq = append(seq, b...)
				out = append(out, seq)
			}
		}
		alts = out
	}
	for len(s) > 0 {
		r, size := utf8.DecodeRuneInString(s)
		switch r {
		case '\\':
			if len(s) < 2 {
				return nil, "", errors.New("glob: trailing backslash")
			}
			r2, size2 := utf8.DecodeRuneInString(s[1:])
			appendNode(node{kind: nText, text: string(r2)})
			s = s[1+size2:]
		case '*':
			s = s[1:]
			for strings.HasPrefix(s, "*") {
				s = s[1:]
			}
			appendNode(node{kind: nSuper})
		case '?':
			appendNode(node{kind: nAny})
			s = s[1:]
		case '[':
			n, rest, err := parseClass(s[1:])
			if err != nil {
				return nil, "", err
			}
			appendNode(n)
			s = rest
		case '{':
			s = s[1:]
			var sub [][]node
			for {
				a, rest, err := parseSeq(s, true)
				if err != nil {
					return nil, "", err
				}
				sub = append(sub, a...)
				if strings.HasPrefix(rest, ",") {
					s = rest[1:]
					continue
				}
				if strings.HasPrefix(rest, "}") {
					s = rest[1:]
					break
				}
				return nil, "", errors.New("glob: unclosed {")
			}
			appendAlts(sub)
		case ',', '}':
			if inBrace {
				return mergeText(alts), s, nil
			}
			appendNode(node{kind: nText, text: string(r)})
			s = s[size:]
		default:
			appendNode(node{kind: nText, text: string(r)})
			s = s[size:]
		}
	}
	if inBrace {
		return nil, "", errors.New("glob: unclosed {")
	}
	return mergeText(alts), "", nil
}

// mergeText joins adjacent text nodes.
func mergeText(alts [][]node) [][]node {
	for i, seq := range alts {
		var out []node
		for _, n := range seq {
			if n.kind == nText && len(out) > 0 && out[len(out)-1].kind == nText {
				out[len(out)-1].text += n.text
				continue
			}
			out = append(out, n)
		}
		alts[i] = out
	}
	return alts
}

// parseClass parses the body of [...] after the opening bracket.
func parseClass(s string) (node, string, error) {
	n := node{kind: nClass}
	if strings.HasPrefix(s, "!") {
		n.not = true
		s = s[1:]
	}
	first := true
	for {
		if s == "" {
			return node{}, "", errors.New("glob: unclosed [")
		}
		r, size := utf8.DecodeRuneInString(s)
		if r == ']' && !first {
			return n, s[size:], nil
		}
		first = false
		if r == '\\' && len(s) > size {
			r, size = utf8.DecodeRuneInString(s[size:])
			size++
		}
		s = s[size:]
		lo, hi := r, r
		if strings.HasPrefix(s, "-") && len(s) > 1 && s[1] != ']' {
			r2, size2 := utf8.DecodeRuneInString(s[1:])
			if r2 == '\\' && len(s) > 1+size2 {
				r2, size2 = utf8.DecodeRuneInString(s[1+size2:])
				size2++
			}
			hi = r2
			s = s[1+size2:]
		}
		if hi < lo {
			return node{}, "", fmt.Errorf("glob: bad range %c-%c", lo, hi)
		}
		n.ranges = append(n.ranges, [2]rune{lo, hi})
	}
}

func (n *node) classMatch(r rune) bool {
	in := false
	for _, rg := range n.ranges {
		if r >= rg[0] && r <= rg[1] {
			in = true
			break
		}
	}
	return in != n.not
}

func (p *globPattern) match(text string) bool {
	for _, seq := range p.alts {
		if matchSeq(seq, text) {
			return true
		}
	}
	return false
}

// matchSeq matches one flat sequence with memoised backtracking.
func matchSeq(seq []node, text string) bool {
	type key struct{ i, pos int }
	memo := map[key]bool{}
	var m func(i, pos int) bool
	m = func(i, pos int) bool {
		k := key{i, pos}
		if v, ok := memo[k]; ok {
			return v
		}
		var res bool
		switch {
		case i == len(seq):
			res = pos == len(text)
		default:
			n := seq[i]
			switch n.kind {
			case nText:
				res = strings.HasPrefix(text[pos:], n.text) && m(i+1, pos+len(n.text))
			case nAny:
				if pos < len(text) {
					_, size := utf8.DecodeRuneInString(text[pos:])
					res = m(i+1, pos+size)
				}
			case nClass:
				if pos < len(text) {
					r, size := utf8.DecodeRuneInString(text[pos:])
					res = n.classMatch(r) && m(i+1, pos+size)
				}
			case nSuper:
				for p := pos; p <= len(text); {
					if m(i+1, p) {
						res = true
						break
					}
					if p == len(text) {
						break
					}
					_, size := utf8.DecodeRuneInString(text[p:])
					p += size
				}
			}
		}
		memo[k] = res
		return res
	}
	return m(0, 0)
}

File: hallpass-py/src/hallpass/integrations/argocd/rbac/glob.py
"""A matcher compatible with github.com/gobwas/glob compiled with no
separators, which is how Argo CD calls it. With no separators ``*`` and
``**`` both match any sequence of characters, including "/".

Syntax: ``*`` and ``**`` any sequence; ``?`` one character; ``[abc]``,
``[a-z]``, ``[!abc]`` character classes; ``{a,b}`` alternatives, which may
nest; ``\\x`` escapes x.
"""

from __future__ import annotations

import threading
from dataclasses import dataclass, field

from hallpass.core.errors import go_quote

__all__ = ["GlobError", "GlobPattern", "compile_glob", "glob_match", "parse_glob"]

_TEXT, _ANY, _SUPER, _CLASS = 0, 1, 2, 3


class GlobError(ValueError):
    pass


@dataclass
class _Node:
    kind: int
    text: str = ""
    # class
    negate: bool = False
    ranges: list[tuple[str, str]] = field(default_factory=list)

    def class_match(self, r: str) -> bool:
        inside = any(lo <= r <= hi for lo, hi in self.ranges)
        return inside != self.negate


@dataclass
class GlobPattern:
    """A compiled pattern: a set of alternative flat sequences."""

    alts: list[list[_Node]]

    def match(self, text: str) -> bool:
        return any(_match_seq(seq, text) for seq in self.alts)


_lock = threading.Lock()
_cache: dict[str, GlobPattern] = {}
_errs: dict[str, GlobError] = {}


def glob_match(pattern: str, text: str) -> bool:
    """Whether text matches pattern. An invalid pattern never matches, as in
    Argo CD (which logs the compile error and returns false)."""
    try:
        p = compile_glob(pattern)
    except GlobError:
        return False
    return p.match(text)


def compile_glob(pattern: str) -> GlobPattern:
    global _cache, _errs
    with _lock:
        p = _cache.get(pattern)
        if p is not None:
            return p
        e = _errs.get(pattern)
        if e is not None:
            raise e
        if len(_cache) + len(_errs) > 10000:
            _cache = {}
            _errs = {}
        try:
            p = parse_glob(pattern)
        except GlobError as err:
            _errs[pattern] = err
            raise
        _cache[pattern] = p
        return p


def parse_glob(pattern: str) -> GlobPattern:
    """Parse the whole pattern. Raises GlobError."""
    alts, rest = _parse_seq(pattern, False)
    if rest != "":
        raise GlobError(f"glob {go_quote(pattern)}: unexpected {go_quote(rest)}")
    return GlobPattern(alts)


def _parse_seq(s: str, in_brace: bool) -> tuple[list[list[_Node]], str]:
    """Parse until end of input or, inside braces, until "," or "}".
    Return every alternative expansion of the parsed sequence."""
    alts: list[list[_Node]] = [[]]

    def append_node(n: _Node) -> None:
        for a in alts:
            a.append(n)

    while s:
        r = s[0]
        if r == "\\":
            if len(s) < 2:
                raise GlobError("glob: trailing backslash")
            append_node(_Node(_TEXT, s[1]))
            s = s[2:]
        elif r == "*":
            s = s[1:].lstrip("*")
            append_node(_Node(_SUPER))
        elif r == "?":
            append_node(_Node(_ANY))
            s = s[1:]
        elif r == "[":
            n, s = _parse_class(s[1:])
            append_node(n)
        elif r == "{":
            s = s[1:]
            sub: list[list[_Node]] = []
            while True:
                a, rest = _parse_seq(s, True)
                sub.extend(a)
                if rest.startswith(","):
                    s = rest[1:]
                    continue
                if rest.startswith("}"):
                    s = rest[1:]
                    break
                raise GlobError("glob: unclosed {")
            alts = [x + y for x in alts for y in sub]
        elif r in ",}":
            if in_brace:
                return _merge_text(alts), s
            append_node(_Node(_TEXT, r))
            s = s[1:]
        else:
            append_node(_Node(_TEXT, r))
            s = s[1:]
    if in_brace:
        raise GlobError("glob: unclosed {")
    return _merge_text(alts), ""


def _merge_text(alts: list[list[_Node]]) -> list[list[_Node]]:
    """Join adjacent text nodes."""
    out_alts = []
    for seq in alts:
        out: list[_Node] = []
        for n in seq:
            if n.kind == _TEXT and out and out[-1].kind == _TEXT:
                out[-1] = _Node(_TEXT, out[-1].text + n.text)
                continue
            out.append(n)
        out_alts.append(out)
    return out_alts


def _parse_class(s: str) -> tuple[_Node, str]:
    """Parse the body of [...] after the opening bracket."""
    n = _Node(_CLASS)
    if s.startswith("!"):
        n.negate = True
        s = s[1:]
    first = True
    while True:
        if s == "":
            raise GlobError("glob: unclosed [")
        r = s[0]
        size = 1
        if r == "]" and not first:
            return n, s[1:]
        first = False
        if r == "\\" and len(s) > 1:
            r = s[1]
            size = 2
        s = s[size:]
        lo = hi = r
        if s.startswith("-") and len(s) > 1 and s[1] != "]":
            r2 = s[1]
            size2 = 1
            if r2 == "\\" and len(s) > 2:
                r2 = s[2]
                size2 = 2
            hi = r2
            s = s[1 + size2 :]
        if hi < lo:
            raise GlobError(f"glob: bad range {lo}-{hi}")
        n.ranges.append((lo, hi))


def _match_seq(seq: list[_Node], text: str) -> bool:
    """Match one flat sequence. The set of text positions reachable after
    each node is tracked, so matching is linear in len(seq) * len(text)
    (Go: memoised backtracking, the same answers)."""
    n = len(text)
    states = {0}
    for node in seq:
        if not states:
            return False
        nxt: set[int] = set()
        if node.kind == _SUPER:
            nxt = set(range(min(states), n + 1))
        elif node.kind == _TEXT:
            for pos in states:
                if text.startswith(node.text, pos):
                    nxt.add(pos + len(node.text))
        elif node.kind == _ANY:
            nxt = {pos + 1 for pos in states if pos < n}
        else:
            nxt = {pos + 1 for pos in states if pos < n and node.class_match(text[pos])}
        states = nxt
    return n in states
